### PR TITLE
SONAR-9025 Optimize computation issue exclusions

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: sonar-scanner-engine/src/test/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsLoaderTest.java [hunk 1]
```diff
@@ -56,27 +50,17 @@
   public TemporaryFolder temp = new TemporaryFolder();
```
Comment (SonarTech) [inline sonar-scanner-engine/src/test/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsLoaderTest.java]: MINOR Remove this unused "TemporaryFolder". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java [hunk 2]
```diff
@@ -116,14 +83,33 @@ public void scan(String resource, Path filePath, Charset encoding) throws IOExce
...
     }
   }
 
+  private void processLine(String line) {
+    if (line.trim().length() == 0) {
+      return;
+    }
+
+    // first check the single regexp patterns that can be used to totally exclude a file
+    for (java.util.regex.Pattern pattern : allFilePatterns) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: MINOR Replace this fully qualified name with "Pattern" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java [hunk 3]
```diff
@@ -19,93 +19,60 @@
...
-      allFilePatterns.add(java.util.regex.Pattern.compile(pattern.getAllFileRegexp()));
-    }
-    for (IssuePattern pattern : patternsInitializer.getBlockPatterns()) {
-      blockMatchers.add(new DoubleRegexpMatcher(
-        java.util.regex.Pattern.compile(pattern.getBeginBlockRegexp()),
-        java.util.regex.Pattern.compile(pattern.getEndBlockRegexp())));
-    }
-
-    init();
+  IssueExclusionsRegexpScanner(String componentKey, List<java.util.regex.Pattern> allFilePatterns, List<DoubleRegexpMatcher> blockMatchers, PatternMatcher patternMatcher) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: MINOR Replace this fully qualified name with "Pattern" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [IssueExclusionsRegexpScanner.java#L144](https://github.com/SonarSource/sonarqube/blob/aa236d99c2f87b8a9ead740fa72470793662f494/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java#L144): Make this a "static" inner class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694)
1. ![MINOR][MINOR] [LineRange.java#L87](https://github.com/SonarSource/sonarqube/blob/aa236d99c2f87b8a9ead740fa72470793662f494/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/LineRange.java#L87): Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [LineRange.java#L97](https://github.com/SonarSource/sonarqube/blob/aa236d99c2f87b8a9ead740fa72470793662f494/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/LineRange.java#L97): Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/PatternDecoder.java [hunk 4]
```diff
@@ -19,100 +19,64 @@
...
-    }
-    return patterns;
+  private PatternDecoder() {
+    // static only
   }
-
-  /**
-   * Main method that decodes a line which defines a pattern
-   */
-  public IssuePattern decodeLine(String line) {
```
Comment (dbmeneses) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/PatternDecoder.java]: @henryju  this is the code I was talking about

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LineRange.java#L87](https://github.com/SonarSource/sonarqube/blob/fa8bbee1cb47ed77cda248e01dfd1a242bd9640b/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/LineRange.java#L87): Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 5]
```diff
@@ -21,20 +21,38 @@
...
 
+import java.io.BufferedInputStream;
+import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.ByteBuffer;
+import java.nio.charset.CharacterCodingException;
 import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.CharsetDecoder'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 6]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.charset.CharsetEncoder;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.CharsetEncoder'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 7]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.charset.CodingErrorAction;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.CodingErrorAction'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 8]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.charset.MalformedInputException;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.MalformedInputException'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 9]
```diff
@@ -21,20 +21,38 @@
...
 import java.nio.charset.StandardCharsets;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.StandardCharsets'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 10]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.charset.UnmappableCharacterException;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.UnmappableCharacterException'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 11]
```diff
@@ -21,20 +21,38 @@
...
 import java.nio.file.Files;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.file.Files'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 12]
```diff
@@ -21,20 +21,38 @@
...
 import java.nio.file.Path;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.file.Path'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 13]
```diff
@@ -21,20 +21,38 @@
...
+import java.util.LinkedHashSet;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.util.LinkedHashSet'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 14]
```diff
@@ -21,20 +21,38 @@
...
+import java.util.Set;
 
+import javax.annotation.CheckForNull;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'javax.annotation.CheckForNull'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 15]
```diff
@@ -21,20 +21,38 @@
...
+
+import org.apache.commons.io.ByteOrderMark;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'org.apache.commons.io.ByteOrderMark'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 16]
```diff
@@ -21,20 +21,38 @@
...
+import org.apache.commons.io.IOUtils;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'org.apache.commons.io.IOUtils'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 17]
```diff
@@ -21,20 +21,38 @@
...
+import org.apache.commons.io.input.BOMInputStream;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'org.apache.commons.io.input.BOMInputStream'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 18]
```diff
@@ -21,20 +21,38 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
+import java.io.BufferedInputStream;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.io.BufferedInputStream'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 19]
```diff
@@ -21,20 +21,38 @@
...
+import java.io.FileInputStream;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.io.FileInputStream'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 20]
```diff
@@ -21,20 +21,38 @@
...
 import java.io.IOException;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.io.IOException'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 21]
```diff
@@ -21,20 +21,38 @@
...
 import java.io.InputStream;
+import java.io.InputStreamReader;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.io.InputStreamReader'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 22]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.ByteBuffer;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.ByteBuffer'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 23]
```diff
@@ -21,20 +21,38 @@
...
+import java.nio.charset.CharacterCodingException;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MINOR Remove this unused import 'java.nio.charset.CharacterCodingException'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  private static boolean tryDecode(byte[] bytes, int len, Charset charset) throws IOException {
+    CharsetDecoder decoder = charset.newDecoder()
+      .onMalformedInput(CodingErrorAction.REPORT)
+      .onUnmappableCharacter(CodingErrorAction.REPORT);
+
+    try {
+      decoder.decode(ByteBuffer.wrap(bytes, 0, len));
+    } catch (CharacterCodingException e) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+    assertRun();
+    return detectedCharset;
+  }
+
+  public InputStream inputStream() {
+    assertRun();
+    return stream;
+  }
+
+  private BOMInputStream createInputStream(Path path) throws IOException {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java]: MAJOR Make "createInputStream" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 22 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 20 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LineRange.java#L87](https://github.com/SonarSource/sonarqube/blob/aa9e3520b268510ef3572244caa5fc83805a2b95/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/LineRange.java#L87): Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: sonar-scanner-engine/src/test/resources/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScannerTest/file-with-single-regexp.txt [hunk 26]
```diff
@@ -30,4 +27,5 @@ public class LineRange {
     return lines;
   }
 
-}
\ No newline at end of file
+}
+// @SONAR-IGNORE-ALL
```
Comment (henryju) [inline sonar-scanner-engine/src/test/resources/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScannerTest/file-with-single-regexp.txt]: Don't you mix file name? Isn't this one supposed to be 'file-with-single-regexp-last-line.txt' ?

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LineRange.java#L87](https://github.com/SonarSource/sonarqube/blob/236af85754800e53cd4aaecae8cf3e898076ee09/sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/pattern/LineRange.java#L87): Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: sonar-scanner-engine/src/test/resources/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScannerTest/file-with-single-regexp.txt (hunk 26 above)
Comment (dbmeneses) [inline sonar-scanner-engine/src/test/resources/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScannerTest/file-with-single-regexp.txt]: yes, it was fixed

**Pull request merged**